std/algebra/emulated/sw_bn254/pairing.go

```diff
@@ -559,16 +558,65 @@ func (pr Pairing) doubleAndAddStep(p1, p2 *g2AffP) (*g2AffP, *lineEvaluation, *l
 	line1.R1 = *pr.Ext2.Sub(&line1.R1, &p1.Y)
 
 	// compute λ2 = -λ1-2y1/(x3-x1)
-	n = pr.Ext2.Double(&p1.Y)
+	n = pr.Ext2.MulByConstElement(&p1.Y, big.NewInt(2))
 	d = pr.Ext2.Sub(x3, &p1.X)
 	l2 := pr.Ext2.DivUnchecked(n, d)
 	l2 = pr.Ext2.Add(l2, l1)
 	l2 = pr.Ext2.Neg(l2)
 
 	// compute x4 = λ2²-x1-x3
 	x4 := pr.Ext2.Square(l2)
-	x4 = pr.Ext2.Sub(x4, &p1.X)
-	x4 = pr.Ext2.Sub(x4, x3)
+	x4 = pr.Ext2.Sub(x4, pr.Ext2.Add(&p1.X, x3))
+
+	// compute y4 = λ2(x1 - x4)-y1
+	y4 := pr.Ext2.Sub(&p1.X, x4)
+	y4 = pr.Ext2.Mul(l2, y4)
+	y4 = pr.Ext2.Sub(y4, &p1.Y)
+
+	p.X = *x4
+	p.Y = *y4
+
+	// compute line2
+	line2.R0 = *l2
+	line2.R1 = *pr.Ext2.Mul(l2, &p1.X)
+	line2.R1 = *pr.Ext2.Sub(&line2.R1, &p1.Y)
+
+	return &p, &line1, &line2
+}
+
+// doubleAndSubStep doubles p1 and subs p2 to the result in affine coordinates, and evaluates the line in Miller loop
+// https://eprint.iacr.org/2022/1162 (Section 6.1)
+func (pr Pairing) doubleAndSubStep(p1, p2 *g2AffP) (*g2AffP, *lineEvaluation, *lineEvaluation) {
```

Comment:
`doubleAndAddStep` and `doubleAndSubStep` are very similar to each other. Instead of two implementations I would rather have a boolean argument `isSub` and then add/subtract based on that.